### ISSUE TITLE
Add property for viewport-aligned images in formatted expressions

### ISSUE THREAD

**Issue by ZeLonewolf** [maplibre/maplibre-gl-js#188]
## Motivation

Rendering of highway shields is a high priority in the US mapping community.  This issue is discussed at length in gravitystorm/openstreetmap-carto#508, where the importance of pictoral shields in North American maps is discussed.  One of the challenges in rendering shields is the rendering of concurrent routes, in which a road shares multiple multiple routes.  There are a few general strategies for this, as discussed in ZeLonewolf/openstreetmap-americana#6.

Of particular interest is the ability to render concurrent routes such that the highway shields are arranged along the run of a route, centered on the rendered road.  In order to achieve this effect, a vector tile server must generate tiles which have all of the desired route concurrencies rendered.  This issue is described in openmaptiles/openmaptiles#1128, where code is in development to expose route concurrencies.

Once route concurrencies are exposed, the routes and refs can be packed into custom image ids and concatenated into a formatted expression containing the images.  A JavaScript handler is then hooked up in order to extract the ref and network from the custom image ids in order to render route numbers on shield blanks.  This approach is also described in ZeLonewolf/openstreetmap-americana#6.

The result of this combination of exposed route concurrencies in OpenMapTiles, image ID packing, and JavaScript shield blank drawing, looks like this (for a particularly long concurrency in Indiana, USA):

![image](https://user-images.githubusercontent.com/3254090/123719457-96443300-d84f-11eb-80d4-78cca5a37203.png)

Note that these shields are rotated along the baseline of the road (as you would expect for a text label).  For a road that runs roughly east/west, this is a reasonable rendering.  However, for north/south roads, the shields are drawn sideways, which is not desirable:

![image](https://user-images.githubusercontent.com/3254090/123719664-1cf91000-d850-11eb-8edc-2a64cdcfe99f.png)

Ideally, there should be an option to suppress line-aligned rotation of images in formatted expressions, so that they can be viewport-aligned for highway shield use cases.

## Design Alternatives

In the specific case of highway shields, one alternative is to render a grid of shields instead of arranging the shields along the route of the highway.  However, this takes up more space and is less aesthetically pleasing.  This is the approach used by Kevin Kenny's mapnik-based [osm-shields project](https://github.com/kennykb/osm-shields/):
![image](https://user-images.githubusercontent.com/3254090/123721497-782d0180-d854-11eb-8112-13fd3b90b200.png)

The approach taken by openstreetmap-carto is to combine the refs into a text label box, however, this approach is unsatisfying to American users, which expect to see shields rendered on highways:
![image](https://user-images.githubusercontent.com/3254090/123721149-acec8900-d853-11eb-8a3d-2a209bf0afa3.png)

Yet another option is to combine multiple refs onto a single shield backdrop, as is done in this paper map.  However, this approach does not scale to cases where multiple route networks are combined in a concurrency:
![image](https://user-images.githubusercontent.com/3254090/123721263-ed4c0700-d853-11eb-8815-dd7d4f189245.png)

## Design

Ideally, a style author should be able to set a property that sets viewport-alignment rotation to images in formatted expressions, while maintaining their position along the LineString.

**Comment by ZeLonewolf** [comment]
Here is an example from Apple Maps of a grouped highway concurrency rendering with upright shields (the effect we're trying to achieve):
![image](https://user-images.githubusercontent.com/3254090/123785474-91f73480-d8a6-11eb-9c62-b554958f7edc.png)

**Comment by wipfli** [comment]
Thanks for the great example pictures. It is clear what the goal is. So route concurrencies are like different names for the same route?

**Comment by wipfli** [comment]
I guess what would be helpful for the implementation in maplibre gl js is a minimal example of code together with some data. Maybe you could provide a style file plus an `.mbtiles` file which shows the shields rotated along the line, i.e. which has the wrong behavior. Then we can elaborate from there how the shield rotation can be fixed.

**Comment by ZeLonewolf** [comment]
A [route concurrency](https://en.wikipedia.org/wiki/Concurrency_(road)) is when multiple routes share a single road.  When long distance routes cross in similar directions, it is common for them to briefly share the same roadway.  Also, concurrencies are common in the case of geographic choke points, such as river crossings, where multiple routes might cross a single bridge.

Below are some examples of route concurrencies from US paper maps:
<img src="https://user-images.githubusercontent.com/3254090/124468462-21995900-dd67-11eb-8eda-3e6991b386a0.jpg" height=300/><img src="https://user-images.githubusercontent.com/3254090/124468465-2231ef80-dd67-11eb-8336-89b52f4e347b.jpg" height=300/><img src="https://user-images.githubusercontent.com/3254090/124468466-2231ef80-dd67-11eb-9826-fffe1fdbfada.jpg" height=300/><img src="https://user-images.githubusercontent.com/3254090/124468467-22ca8600-dd67-11eb-9e62-e92f4e219131.jpg" height=300/>

A test case for this functionality is provided in two files:
1. [tiles.mbtiles.tar.gz](https://github.com/maplibre/maplibre-gl-js/files/6764057/tiles.mbtiles.tar.gz): a tiles.mbtiles containing vector tile highway data for Indiana, USA
2. [highway-shield-test-style.tar.gz](https://github.com/maplibre/maplibre-gl-js/files/6764060/highway-shield-test-style.tar.gz): an executable example style demonstrating shield rendering

The style uses javascript callbacks in order to dynamically generate the shield numbers

The style can be run using "make run" (requires python3) and will run on port 1776.  A suggested test area on the map is: http://localhost:1776/#10/39.8086/-86.109

The relevant code for generating the shield ids in formatted text is in `layer/road.js`, specifically this block:

```
    "text-field": [
      "format",
      [
        "case",
        ["!=", ["get", "ref_1"], null],
        [
          "image",
          ["concat", "shield_", ["get", "network_1"], "=", ["get", "ref_1"]],
        ],
        ["literal", ""],
      ],
      " ",
      [
        "case",
        ["!=", ["get", "ref_2"], null],
        [
          "image",
          ["concat", "shield_", ["get", "network_2"], "=", ["get", "ref_2"]],
        ],
        ["literal", ""],
      ],
      " ",
      [
        "case",
        ["!=", ["get", "ref_3"], null],
        [
          "image",
          ["concat", "shield_", ["get", "network_3"], "=", ["get", "ref_3"]],
        ],
        ["literal", ""],
      ],
      " ",
      [
        "case",
        ["!=", ["get", "ref_4"], null],
        [
          "image",
          ["concat", "shield_", ["get", "network_4"], "=", ["get", "ref_4"]],
        ],
        ["literal", ""],
      ],
      " ",
      [
        "case",
        ["!=", ["get", "ref_5"], null],
        [
          "image",
          ["concat", "shield_", ["get", "network_5"], "=", ["get", "ref_5"]],
        ],
        ["literal", ""],
      ],
      " ",
      [
        "case",
        ["!=", ["get", "ref_6"], null],
        [
          "image",
          ["concat", "shield_", ["get", "network_5"], "=", ["get", "ref_6"]],
        ],
        ["literal", ""],
      ],
    ],
```
Note that the schema used in the .mbtiles is not stock OpenMapTiles; it's a development version of openmaptiles/openmaptiles#1140, which creates a new `transportation_ref` layer dedicated to route rendering.

**Comment by wipfli** [comment]
Thanks Brian for the files. I get this in the console:

```
➜  highway-shield-concurrency-style make run
python3 -m http.server 1776
Serving HTTP on 0.0.0.0 port 1776 (http://0.0.0.0:1776/) ...
***.***.***.*** - - [08/Jul/2021 20:38:44] "GET /layer/tunnel.js HTTP/1.1" 200 -
***.***.***.*** - - [08/Jul/2021 20:38:44] "GET /layer/oneway.js HTTP/1.1" 200 -
***.***.***.*** - - [08/Jul/2021 20:38:44] code 404, message File not found
***.***.***.*** - - [08/Jul/2021 20:38:44] "GET /layer/highway_exit.js HTTP/1.1" 404 -
***.***.***.*** - - [08/Jul/2021 20:38:44] code 404, message File not found
***.***.***.*** - - [08/Jul/2021 20:38:44] "GET /layer/highway_exit.js HTTP/1.1" 404 -
```

and the browser does not show anything. Can you maybe package this code in a GitHub repo such that we can work together on it?

**Comment by ZeLonewolf** [comment]
I realize I've left out a key piece in my writeup - the style is pointing to a local openmaptiles server, so it's expecting the tarred up .mbtiles to be hosted on a local tileserver running on port 8080.  In any case, the 404s indicate that I goofed up in the attachment, so let me work on the suggested GH repo.

**Comment by wipfli** [comment]
Great, thanks!

**Comment by ZeLonewolf** [comment]
I've built up a github sample project that has the running sample:

https://github.com/ZeLonewolf/maplibre-shield-rotation-sample

**Comment by ZeLonewolf** [comment]
This issue remains a blocker for American-style map rendering, as discussed in ZeLonewolf/openstreetmap-americana#6.

**Comment by wipfli** [comment]
#716

**Comment by JannikGM** [comment]
Should have been closed by #716, or is there anything left to do?

**Comment by HarelM** [comment]
Yup. Resolved by #716
